src/pkcs11/mechanism.c

```diff
@@ -38,11 +38,51 @@ struct hash_signature_info {
 struct signature_data {
 	struct sc_pkcs11_object *key;
 	struct hash_signature_info *info;
-	sc_pkcs11_operation_t *	md;
-	CK_BYTE			buffer[4096/8];
+	sc_pkcs11_operation_t *md;
+	CK_BYTE			*buffer;
 	unsigned int	buffer_len;
 };
 
+static struct signature_data *
+new_signature_data()
+{
+	return calloc(1, sizeof(struct signature_data));
+}
+
+static void
+signature_data_release(struct signature_data *data)
+{
+	if (!data)
+		return;
+	sc_pkcs11_release_operation(&data->md);
+	sc_mem_secure_free(data->buffer, data->buffer_len);
```

Comment:
would it make sense to do `sc_mem_clear()` here too as it was in the previous implementation or using the secure memory implies this somehow (might be worth mentioning in comment?)?